src/mo_frontend/typing.ml

```diff
@@ -1831,11 +1835,20 @@ and check_exp' env0 t exp : T.typ =
     check_cases env t1 t cases;
     coverage_cases "switch" env cases t1 exp.at;
     t
-  | TryE (exp1, cases), _ ->
+  | TryE (exp1, cases, exp2_opt), _ ->
     check_ErrorCap env "try" exp.at;
     check_exp env t exp1;
     check_cases env T.catch t cases;
-    coverage_cases "try handler" env cases T.catch exp.at;
+    if cases <> []
+    then coverage_cases "try handler" env cases T.catch exp.at;
+    if not env.pre then
+      begin match exp2_opt with
+      | None -> ()
+      | Some exp2 ->
+        check_exp_strong { env with rets = None; labs = T.Env.empty } T.unit exp2;
+        if exp2.note.note_eff <> T.Triv then
+          local_error env exp2.at "M0200" "a cleanup clause must not send messages";
+      end;
```

Comment:
I think a better way to prevent this is to set the env.asyncCap to NullCap so we just allow synchronous code here.

Reply:
I'll try!

Reply:
Done in #4622